homeassistant/components/teslemetry/calendar.py

```diff
@@ -0,0 +1,290 @@
+"""Calendar platform for Teslemetry integration."""
+
+from __future__ import annotations
+
+from datetime import datetime, timedelta
+from typing import Any
+
+from attr import dataclass
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+from homeassistant.util import dt as dt_util
+
+from . import TeslemetryConfigEntry
+from .entity import TeslemetryEnergyInfoEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: TeslemetryConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up the Teslemetry Calendar platform from a config entry."""
+
+    entities_to_add: list[CalendarEntity] = []
+
+    # Add buy tariff calendar entities
+    entities_to_add.extend(
+        TeslemetryTariffSchedule(energy, "tariff_content_v2")
+        for energy in entry.runtime_data.energysites
+        if energy.info_coordinator.data.get("tariff_content_v2_seasons")
+    )
+
+    # Add sell tariff calendar entities
+    entities_to_add.extend(
+        TeslemetryTariffSchedule(energy, "tariff_content_v2_sell_tariff")
+        for energy in entry.runtime_data.energysites
+        if energy.info_coordinator.data.get("tariff_content_v2_sell_tariff_seasons")
+    )
+
+    async_add_entities(entities_to_add)
+
+
+@dataclass
+class TariffPeriod:
+    """A single tariff period."""
+
+    name: str
+    price: float
+    from_hour: int = 0
+    from_minute: int = 0
+    to_hour: int = 0
+    to_minute: int = 0
+
+
+class TeslemetryTariffSchedule(TeslemetryEnergyInfoEntity, CalendarEntity):
+    """Energy Site Tariff Schedule Calendar."""
+
+    seasons: dict[str, dict[str, Any]] = {}
+    charges: dict[str, dict[str, Any]] = {}
```

Comment:
Adding mutables in the class declaration means they're shared among all instances of the class.
It doesn't seem to create problems now, because we replace them in `_async_update_attrs`. Still, this is an anti-pattern in Python, please move the initialization to `__init__`